sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/MessageFlux.java

```diff
@@ -0,0 +1,1045 @@
+                drain(null);
+            } else {
+                // Once the drain-loop processed the last error, then further errors dropped through the standard
+                // Reactor channel. E.g., retry exhaust error happened and, as part of its processing, upstream
+                // gets canceled, but if upstream still signals an error, then that error gets dropped.
+                Operators.onErrorDropped(e, messageSubscriber.currentContext());
+            }
+        }
+
+        /**
+         * Signals operator termination with completion.
+         * <ul>
+         *     <li>Invoked by the upstream when it completes, or</li>
+         *     <li>Invoked by the the drain-loop when it detects first receiver signaled terminal completion and retry
+         *     is disabled (i.e., NULL_RETRY_POLICY is set).</li>
+         * </ul>
+         */
+        @Override
+        public void onComplete() {
+            if (done) {
+                return;
+            }
+
+            done = true;
+            final String logMessage;
+            if (retryPolicy == NULL_RETRY_POLICY) {
+                logMessage = "Terminal completion signal from Upstream|Receiver arrived at MessageFlux.";
+            } else {
+                logMessage = "Terminal completion signal from Upstream arrived at MessageFlux.";
+            }
+            mediatorHolder.withReceiverInfo(logger.atWarning()).log(logMessage);
+            drain(null);
+        }
+
+        /**
+         * Invoked by the downstream message subscriber to signal the demand for messages. Whatever has been
+         * requested can be sent downstream, so only signal the demand for what can be safely handled.
+         * No messages will be sent downstream until the demand is signaled.
+         *
+         * @param n the number of messages to send to downstream.
+         */
+        @Override
+        public void request(long n) {
+            if (Operators.validate(n)) {
+                Operators.addCap(REQUESTED, this, n);
+                drain(null);
+            }
+        }
+
+        /**
+         * Invoked by downstream to signal termination by cancellation.
+         */
+        @Override
+        public void cancel() {
+            if (cancelled) {
+                return;
+            }
+
+            cancelled = true;
+            mediatorHolder.withReceiverInfo(logger.atWarning())
+                .log("Downstream cancellation signal arrived at MessageFlux.");
+            // By incrementing wip, indicate the active drain-loop that there is cancel signal to handle,
+            if (WIP.getAndIncrement(this) == 0) {
+                // but it is identified that there is no active drain-loop; hence immediately react to cancel.
+                upstream.cancel();
+                mediatorHolder.freeze();
+                // wip increment also placed a tombstone on the drain-loop, so further drain(..) calls are nop.
```

Comment:
nit: nop -> no-op

Thanks for the documentation. It is a great read.

Reply:
Ha.. there are a few nop text in this class, will rename to no-op.